daml-lf/spec/daml-lf-1.rst

```diff
@@ -1887,140 +1887,123 @@ will always be used to compare values of same types::
-   Mod:T:E₁ <ᵥ Mod:T:E₂
+  ———————————————————————————————————————————————————
+    'Update' <ₜ Mod:T
 
-   v₁ ~ᵥ v₁   ⋯   vᵢ₋₁ ~ᵥ wᵢ₋₁     vᵢ <ᵥ wᵢ    i <= n
-  ——————————————————————————————————————————————————— GenLtStructCon
-   ⟨ f₁ = v₁, …, fₘ = vₘ ⟩ ~ᵥ ⟨ f₁ = w₁, …, fₘ = wₘ ⟩
+    PkgId₁ < PkgId₂
+  ———————————————————————————————————————————————————
+    (PkgId₁:ModName₁):T₁ <ₜ (PkgId₂:ModName₂):T₂
+
+    ModName₁ < ModName₂
+  ———————————————————————————————————————————————————
+    (PkgId:ModName₁):T₁ <ₜ (PkgId:ModName₂):T₂
+
+    T₁ < T₂
+  ———————————————————————————————————————————————————
+    Mod:T₁ <ₜ Mod:T₂
+
+  ———————————————————————————————————————————————————
+    Mod:T <ₜ n
+
+    n₁ is strictly less that n₂
+  ———————————————————————————————————————————————————
+    n₁ <ₜ n₂
+
+  ———————————————————————————————————————————————————
+    n <ₜ ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩
+
+    fᵢ < gᵢ
+  ———————————————————————————————————————————————————
+    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
+      ⟨ f₁ : σ₁, …, fᵢ₋1 : σᵢ₋₁, gᵢ : σᵢ, …, gₙ : σₙ ⟩
 
-  ——————————————————————————————————————————————————— GenLtTextMap1
-   [ ] <ᵥ [s₁ ↦ w₁, …, sₘ ↦ wₘ]
-
-   t₁ < t₂ < … < tₘ
-   s₁ < s₂ < … < sₙ
-   t₁ < s₁
-  ——————————————————————————————————————————————————— GenLtTextMap2
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-   t₁ < t₂ < … < tₘ
-   s₁ < s₂ < … < sₙ
-   t₁ = s₁
-   v₁ <ᵥ w₁
-  ——————————————————————————————————————————————————— GenLtTextMap3
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-   t₁ = s₁
-   v₁ ~ᵥ w₁
-   [t₂ ↦ v₂, …, tₘ ↦ vₘ] <ᵥ [s₂ ↦ w₂, …, sₙ ↦ wₙ]
-  ——————————————————————————————————————————————————— GenLtTextMap4
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-  ——————————————————————————————————————————————————— GenLtGenMap1
-   [ ] <ᵥ [s₁ ↦ w₁, …, sₘ ↦ wₘ]
-
-   t₁ <ᵥ t₂ <ᵥ … <ᵥ tₘ
-   s₁ <ᵥ s₂ <ᵥ … <ᵥ sₙ
-   t₁ <ᵥ s₁
-  ——————————————————————————————————————————————————— GenLtGenMap2
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-   t₁ <ᵥ t₂ <ᵥ … <ᵥ tₘ
-   s₁ <ᵥ s₂ <ᵥ … <ᵥ sₙ
-   t₁ ~ᵥ s₁
-   v₁ <ᵥ w₁
-  ——————————————————————————————————————————————————— GenLtGenMap3
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-   t₁ ~ᵥ s₁
-   v₁ ~ᵥ w₁
-   [t₂ ↦ v₂, …, tₘ ↦ vₘ] <ᵥ [s₂ ↦ w₂, …, sₙ ↦ wₙ]
-  ——————————————————————————————————————————————————— GenLtGenMap4
-   [t₁ ↦ v₁, …, tₘ ↦ vₘ] <ᵥ [s₁ ↦ w₁, …, sₙ ↦ wₙ]
-
-.. note: In the above rules, map entries for TextMap and GenMap are ordered
-   by key. The rules make this assumption explicit.
+  ———————————————————————————————————————————————————
+    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
+      ⟨ f₁ : τ₁, …, fₘ : τₘ, …, fₙ : σₙ ⟩
 
+    τᵢ <ₜ σᵢ
+  ———————————————————————————————————————————————————
+    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ
+      ⟨ f₁ : τ₁, …, fᵢ₋₁ : τᵢ₋₁, fᵢ : σᵢ, …, fₘ : σₘ ⟩
+
+  ———————————————————————————————————————————————————
+    ⟨ f₁ : τ₁, …, fₘ : τₘ ⟩ <ₜ τ σ
+
+    τ₁ <ₜ τ₂
```

Comment:
Maybe I’m missing something but which rules allows me to deduce that `Int64 < Numeric 10`?

Reply:
By transitivity (stated in the text)  you have  ``Int64 <ₜ  τ σ``  for all ``τ`` and ``σ``.

I replace the text by an explicit rule.

Reply:
I don't see it @remyhaemmerle-da . There's no rule that lets you infer, for example, that ` τ₁ <ₜ τ₂ σ` whenever ` τ₁` is not an application type.

Reply:
Yeah I also cannot find it in the text.

Reply:
- Using `TypeOrderTransitivity`, `TypeOrderInt64Date`, and `TypeOrderDateTimestamp` you can infer that 'Int64' <ₜ 'Timestamp'.
- Combining with  `TypeOrderTransitivity`, and  `TypeOrderTimestampText` you can infer that  `'Int64' <ₜ 'Text'`
 ...
- Combining with  `TypeOrderTransitivity`, and  `TypeOrderNatStruct` you can infer that ` 'Int64' <ₜ ⟨  ⟩`
- Combining with  `TypeOrderTransitivity` and `TypeOrderStructTyApp` you can infer that  `'Int64' <ₜ Numeric 10`

Reply:
Oh, I see! That's very hard to see, but it checks out. 

Maybe it would be easier to understand if we used the type rank notion from the scala implementation? We could define a formal type rank function, and thenhave just one rule for comparing all the types with different type rank.

Reply:
I do not like to define the rank explicitly (mapping types to integers)

Reply:
I think it would make the spec match the code more closely, and be easier to understand, which are both better for maintenance, but if you are sure about this way then I'm not opposed.

Reply:
I want to be able to introduce a new type between two already existing.  
This will oblige me to have different ranks between two specs.

Reply:
That's fine. The type rank function is entirely formal, and would only exist in the spec, not be exposed in DAML/DAML-LF. Alternatively, you can define a formal enumeration of type ranks (instead of an type rank integer) and use the enumeration order.

Anyway, it is up to you.

Reply:
I did not include the rank.